keps/sig-cluster-lifecycle/0032-create-a-k8s-io-component-repo.md

```diff
@@ -0,0 +1,245 @@
+- Avoid moving code into _k8s.io/apiserver_ which does not strictly belong to an etcd-based, API group-serving apiserver. Corollary: remove etcd dependency from components.
+- Components should be consistent in how they load (and write) configuration, and merge config with CLI flags.
+- Factor out command- and flag-building code to a shared place.
+- Factor out common HTTPS endpoints describing a component's status.
+- Make the core Kubernetes components utilize these new packages.
+- Have good documentation about how to build a component with a similar look and feel as core Kubernetes components.
+- Increase test coverage for the configuration, command building, and HTTPS server areas of the component code.
+- Break out OpenAPI definitions and violations for the current ComponentConfigs from the core repo to a dedicated place per-component. With auto-generated OpenAPI specs for each component ComponentConfig consumers can validate/vet their configs without running the component.
+
+### Success metrics
+
+- All core Kubernetes components (kube-apiserver, kube-controller-manager, kube-scheduler, kubelet, kube-proxy, kubeadm) are using these shared packages in a consistent manner.
+- Cloud providers can be moved out of core without having to depend on the core repository.
+  - Related issue: [https://github.com/kubernetes/kubernetes/issues/69585](https://github.com/kubernetes/kubernetes/issues/69585)
+- It's easier for _kubeadm_ to move out of the core repo when these component-related packages are in a "public" staging repository.
+
+### Non-goals
+
+- Graduate any ComponentConfig API versions (in this proposal).
+- Make this library toolkit a "generic" cloud-native component builder. Such a toolbox, if ever created, could instead consume these packages. In other words, this repository is solely focused on Kubernetes' components' needs.
+- Fixing _all the problems_ our components have right now, and expanding this beyond what's really necessary. Instead working incrementally, and starting to break out some basic stuff we _know_ every component must handle (e.g. configuration and flag parsing)
+
+### Related proposals / references
+
+- [Kubernetes Component Configuration](https://docs.google.com/document/d/1arP4T9Qkp2SovlJZ_y790sBeiWXDO6SG10pZ_UUU-Lc/edit) by [@mikedanese](https://github.com/mikedanese)
+- [Versioned Component Configuration Files](https://docs.google.com/document/d/1FdaEJUEh091qf5B98HM6_8MS764iXrxxigNIdwHYW9c/edit#) by [@mtaufen](https://github.com/mtaufen)
+- [Moving ComponentConfig API types to staging repos](https://github.com/kubernetes/community/blob/master/keps/sig-cluster-lifecycle/0014-20180707-componentconfig-api-types-to-staging.md) by [@luxas](https://github.com/luxas) and [@sttts](https://github.com/sttts)
+
+## Proposal
+
+This proposal contains three logical units of work. Each subsection is explained in more detail below.
+
+### Part 1: ComponentConfig
+
+#### Standardized encoding/decoding
```

Comment:
For this section, what in `k8s.io/component/config/serializer` would be different from the stuff already in `k8s.io/apimachinery/...`?

Reply:
Convenience helpers, e.g. for the top-1 question asked about apimachinery: "how do I read a yaml file?"

Reply:
Yeah, and that's "harder" than it has to be. With this in place, it'd be one simple fn call

Reply:
I'm still not sure I understand why we don't want to provide these convenience helpers in apimachinery - especially if there's a high demand for them from folks using those libraries. ComponentConfig isn't the only use-case for wanting to read K8s style YAML, for example.

Reply:
> would we put it in this new place instead of extending apimachinery?

Ah sorry, I didn't see this part. Well if we think that level of abstraction is applicable somewhere in `k8s.io/apimachinery` I'm happy with having it there as well